browser/ui/webui/brave_donate_ui.cc

```diff
@@ -0,0 +1,187 @@
+
+namespace {
+
+// The handler for Javascript messages for Brave about: pages
+class RewardsDonateDOMHandler : public WebUIMessageHandler,
+                                public brave_rewards::RewardsServiceObserver {
+public:
+  RewardsDonateDOMHandler() {};
+  ~RewardsDonateDOMHandler() override;
+
+  void Init();
+
+  // WebUIMessageHandler implementation.
+  void RegisterMessages() override;
+
+private:
+  void GetPublisherDonateData(const base::ListValue* args);
+  void GetWalletProperties(const base::ListValue* args);
+
+  // RewardsServiceObserver implementation
+  void OnWalletProperties(brave_rewards::RewardsService* rewards_service,
+      int error_code,
+      std::unique_ptr<brave_rewards::WalletProperties> wallet_properties) override;
+
+  brave_rewards::RewardsService* rewards_service_;  // NOT OWNED
+
+  DISALLOW_COPY_AND_ASSIGN(RewardsDonateDOMHandler);
+};
+
+RewardsDonateDOMHandler::~RewardsDonateDOMHandler() {
+  if (rewards_service_)
+    rewards_service_->RemoveObserver(this);
+}
+
+void RewardsDonateDOMHandler::Init() {
+  Profile* profile = Profile::FromWebUI(web_ui());
+  rewards_service_ =
+      brave_rewards::RewardsServiceFactory::GetForProfile(profile);
+  if (rewards_service_)
+    rewards_service_->AddObserver(this);
+}
+
+void RewardsDonateDOMHandler::RegisterMessages() {
+  web_ui()->RegisterMessageCallback("brave_rewards_donate.getPublisherBanner",
+                                    base::BindRepeating(&RewardsDonateDOMHandler::GetPublisherDonateData,
+                                                        base::Unretained(this)));
+  web_ui()->RegisterMessageCallback("brave_rewards_donate.getWalletProperties",
+                                    base::BindRepeating(&RewardsDonateDOMHandler::GetWalletProperties,
+                                                        base::Unretained(this)));
+}
+
+void RewardsDonateDOMHandler::GetPublisherDonateData(const base::ListValue* args) {
+  if (!web_ui()->CanCallJavascript()) {
+    return;
+  }
+
+  std::string publisher_key;
+  args->GetString(0, &publisher_key);
+  brave_rewards::PublisherBanner banner = rewards_service_->GetPublisherBanner(publisher_key);
+
+  base::DictionaryValue result;
+  result.SetString("publisherKey", publisher_key);
+  result.SetString("title", banner.title);
+  result.SetString("description", banner.description);
+  result.SetString("background", banner.background);
+  result.SetString("logo", banner.logo);
+
+  auto amounts = std::make_unique<base::ListValue>();
+  for (int const& value : banner.amounts) {
+    amounts->AppendInteger(value);
+  }
+  result.SetList("amounts", std::move(amounts));
+
+  auto social = std::make_unique<base::DictionaryValue>();
+  for (auto const& item : banner.social) {
+    social->SetString(item.first, item.second);
+  }
+  result.SetDictionary("social", std::move(social));
+
+  web_ui()->CallJavascriptFunctionUnsafe("brave_rewards_donate.publisherBanner", result);
```

Comment:
- [ ] Use `CallJavascriptFunction` instead?

Reply:
I forgot why, but `CallJavascriptFunction` was not working. cc @bbondy if you can remember

we currently use `CallJavascriptFunctionUnsafe` across all our webui logic